numpy/core/src/umath/ufunc_object.c

```diff
@@ -2219,25 +2264,51 @@ PyUFunc_GeneralizedFunction(PyUFuncObject *ufunc,
     /* Get all the arguments */
     retval = get_ufunc_arguments(ufunc, args, kwds,
                 op, &order, &casting, &extobj,
-                &type_tup, &subok, NULL, &axes);
+                &type_tup, &subok, NULL, &axes, &keepdims);
     if (retval < 0) {
         goto fail;
     }
-
+    /*
+     * If keepdims was passed in (and thus changed from the initial value
+     * on top), check the gufunc is suitable.
+     */
+    if (keepdims != -1) {
+        retval = _check_keepdims_support(ufunc);
+        if (retval < 0) {
+            goto fail;
+        }
+    }
+    /*
+     * If keepdims is set, all input and output core dimensions are the same.
+     */
+    if (keepdims == 1) {
+        for (i = 0; i < nop; ++i) {
+            core_num_dims_array[i] = ufunc->core_num_dims[0];
```

Comment:
minor nit: I'd find this easier to read with an assignment `all_num_dims = ufunc->core_num_dims[0]` or similar outside the loop - too easy to read this as an elementwise copy. But maybe I'm just tired